src/ott/solvers/iterative_linear.py

```diff
@@ -0,0 +1,303 @@
+from ott.solvers.linear import sinkhorn, sinkhorn_lr
+
+__all__ = ["IterativeLinearSolver", "IterLinState"]
+
+LinearOutput = Union[sinkhorn.SinkhornOutput, sinkhorn_lr.LRSinkhornOutput]
+
+ProgressCallbackFn_t = Callable[
+    [Tuple[np.ndarray, np.ndarray, np.ndarray, "IterLinState"]], None]
+
+
+class IterLinState(NamedTuple):
+  """State of the IterativeLinearSolver.
+
+  Attributes:
+  costs: Holds the sequence of costs seen through the outer
+  loop of the solver.
+  linear_convergence: Holds the sequence of bool convergence flags of the
+  inner linear solvers.
+  linear_pb: last inner linear_problem
+  linear_state: solution to the linear_pb
+  prob_state: a problem specific state, provided and managed by
+    user-defined functions
+  rngs: random keys passed to the user defined functions at each iteration.
+  errors: Holds sequence of vectors of errors of the Sinkhorn algorithm
+  at each iteration.
+  """
+
+  costs: jnp.ndarray
+  linear_convergence: jnp.ndarray
+  linear_pb: linear_problem.LinearProblem
+  linear_state: LinearOutput
+  prob_state: NamedTuple = None
+  rngs: Optional[jax.random.PRNGKeyArray] = None
+  errors: Optional[jnp.ndarray] = None
+
+  def set(self, **kwargs: Any) -> "IterLinState":
+    """Return a copy of self, possibly with overwrites."""
+    return self._replace(**kwargs)
+
+  def update(  # noqa: D102
+      self,
+      iteration: int,
+      cost: float,
+      linear_pb: linear_problem.LinearProblem,
+      linear_sol: LinearOutput,
+      prob_state: NamedTuple,
+      store_errors: bool,
+  ) -> "IterLinState":
+    costs = self.costs.at[iteration].set(cost)
+    errors = None
+    if store_errors and self.errors is not None:
+      errors = self.errors.at[iteration, :].set(linear_sol.errors)
+    linear_convergence = self.linear_convergence.at[iteration].set(
+        linear_sol.converged
+    )
+    return self.set(
+        linear_pb=linear_pb,
+        linear_state=linear_sol,
+        prob_state=prob_state,
+        costs=costs,
+        linear_convergence=linear_convergence,
+        errors=errors,
+    )
+
+
+@jax.tree_util.register_pytree_node_class
+class IterativeLinearSolver(was_solver.WassersteinSolver):
+  """Implements probems requiring iterative calls to (low_rank) sinkhorn solver.
+
+  Args:
+  args: Positional arguments for
+  :class:`~ott.solvers.was_solver.WassersteinSolver`.
+  warm_start: Whether to initialize (low-rank) Sinkhorn calls using values
+  from the previous iteration. If `None`, warm starts are not used for
+  standard Sinkhorn, but used for low-rank Sinkhorn.
+  progress_fn: callback function which gets called during the
+  inner iterations, so the user can display the error at each
+  iteration, e.g., using a progress bar.
+  See :func:`~ott.utils.default_progress_fn` for a basic implementation.
+  kwargs_init: auxilary key arguments
+  """
+
+  def __init__(
+      self,
+      *args: Any,
+      warm_start: Optional[bool] = None,
```

Comment:
I don't think any of these arguments should be exposed here (as it currently stands):
- warm_start can be removed
- `progress_fn` - usually is passed to the inner solver in `**kwargs`
- `kwargs_init` - would pass in `__call__` as `**kwargs`